File: spring-boot/src/main/java/com/erdaldalkiran/cart/domain/coupon/Coupon.java
package com.erdaldalkiran.cart.domain.coupon;

import com.erdaldalkiran.cart.domain.cart.Cart;
import com.erdaldalkiran.cart.domain.discount.DiscountType;
import lombok.Getter;

import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.function.BiFunction;

@Getter
public class Coupon {
    private final UUID id;
    private final BigDecimal minimumCartAmount;
    private final DiscountType discountType;
    private final BigDecimal rate;


    public Coupon(UUID id, BigDecimal minimumCartAmount, DiscountType discountType, BigDecimal rate) {
        this.id = id;
        this.minimumCartAmount = minimumCartAmount;
        this.discountType = discountType;
        this.rate = rate;
    }

    public boolean isApplicableTo(Cart cart) {
        var amount = cart.getTotalAmountAfterCampaign();
        return amount.compareTo(minimumCartAmount) > 0;
    }

    public Optional<BigDecimal> calculateDiscountAmountFor(Cart cart) {
        var isApplicable = isApplicableTo(cart);
        if (!isApplicable) return Optional.empty();

        var cartAmount = cart.getTotalAmountAfterCampaign();
        var discountAmount = CouponDiscountAmountCalculator.Strategies
                .get(this.discountType).apply(cartAmount, this.rate);

        var  min = discountAmount.min(cartAmount);

        return Optional.of(min);
    }

    static class CouponDiscountAmountCalculator {
        static Map<DiscountType, BiFunction<BigDecimal,BigDecimal, BigDecimal>> Strategies = Map.of(
                DiscountType.Amount, (amount, rate) -> {
                    var discountAmount = rate;
                    return discountAmount.round(new MathContext(2, RoundingMode.HALF_EVEN));
                },
                DiscountType.Rate, (amount, rate) -> {
                    var discountAmount = amount.multiply(rate);
                    return discountAmount.round(new MathContext(2, RoundingMode.HALF_EVEN));
                }
        );
    }
}

File: dotnet/src/ShoppingCart.Business/Coupon/Coupon.cs
﻿using System;
using System.Collections.Generic;
using ShoppingCart.Business.Campaign;

namespace ShoppingCart.Business.Coupon
{
    public class Coupon
    {
        public Guid ID { get; }

        public decimal MinimumCartAmount { get; }

        public DiscountType Type { get; }

        public decimal Rate { get; }

        public Coupon(Guid id, decimal minimumCartAmount, DiscountType type, decimal rate)
        {
            ID = id;
            MinimumCartAmount = Math.Round(minimumCartAmount, 2);
            Type = type;
            Rate = rate;
        }

        public bool IsApplicableTo(Cart.Cart cart)
        {
            var cartAmount = cart.TotalAmountAfterCampaign;

            return cartAmount > MinimumCartAmount;
        }

        public decimal? CalculateDiscountAmountFor(Cart.Cart cart)
        {
            var isApplicable = IsApplicableTo(cart);
            if (!isApplicable) return null;

            var cartAmount = cart.TotalAmountAfterCampaign;

            return Math.Min(cart.TotalAmountAfterCampaign, CouponDiscountAmountCalculator.Strategies[Type](cartAmount, Rate));
        }
    }

    public static class CouponDiscountAmountCalculator
    {
        public static Dictionary<DiscountType, Func<decimal, decimal, decimal>> Strategies =
            new Dictionary<DiscountType, Func<decimal, decimal, decimal>>
            {
                {
                    DiscountType.Amount, (amount, rate) =>
                    {
                        var discountAmount = rate;
                        return Math.Round(discountAmount, 2);
                    }
                },
                {
                    DiscountType.Rate, (amount, rate) =>
                    {
                        var discountAmount = amount * rate;
                        return Math.Round(discountAmount, 2);
                    }
                }
            };
    }
}